sopel: move deprecated() to sopel.lifecycle

Exirel opened the pull request:
### Description

The `deprecated()` function is used everywhere, whenever we need to deprecate something. Sometimes, that somethings is right where this function is defined: in `sopel.tools`, causing cyclic import errors.

In an effort to reduce cyclic import errors, and to organize the code a bit better (mostly by removing everything from `sopel/tools/__init__.py`), this function is moved to its own sopel's submodule: `sopel.lifecycle`.

Documentation has been updated accordingly, and every part of Sopel that used `sopel.tools.deprecated` is now using `sopel.lifecycle.deprecated`.

In order to prevent yet another cyclic import error, `deprecated` no longer uses `sopel.tools.get_logger` to get the logger for an external plugins (that isn't using the namespace), but tries to mimic its behavior for namespace plugins only.

**Edit**: After discussion, I moved from `sopel.deprecated` to `sopel.lifecycle`. This module can be used, in the **future** to manage more version & lifecycle related things. But for now, I'd rather see this merged than trying to do too much and be frustrated because I've other things in mind that are blocked by this change.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

dgw commented on sopel/deprecated.py:
```diff
@@ -0,0 +1,178 @@
+                    # core, or core plugin
+                    logger = logging.getLogger(module_name)
+                else:
+                    # probably a plugin; use Sopel's public API for getting the
+                    # logger for a plugin
```
Not true any more. This just uses `logging` directly.

```suggestion
                    # probably a plugin; try to handle most cases sanely
```

dgw commented on sopel/deprecated.py:
```diff
@@ -0,0 +1,178 @@
+
+    .. versionadded:: 7.1
+        The ``warning_in`` and ``stack_frame`` parameters.
+
```
For total completeness, one could add this:

```suggestion

.. versionchanged:: 8.0
    Moved out of :mod:`sopel.tools` to solve circular dependency issues.

```

dgw commented on docs/source/deprecated.rst:
```diff
@@ -0,0 +1,32 @@
+Deprecated
+==========
+
+As Sopel grows and evolves, it needs to say goodbye to obsolete pieces of code
+and to replace them with the appropriate solution: sometimes with new code,
+sometimes with new dependencies, sometimes by putting the feature into an
+external plugins.
```
```suggestion
As Sopel grows and evolves, sometimes it needs to say goodbye to obsolete code
and replace it. The solution can be new (and hopefully better) code, switching
to an external library, or maybe putting the feature into an external plugin.
```

dgw commented on docs/source/deprecated.rst:
```diff
@@ -0,0 +1,32 @@
+==========
+Deprecated
+==========
```
```suggestion
============
Deprecations
============
```

\+ rename the file

I'm wondering if the module wouldn't be better named `sopel.deprecation`, actually.

dgw replied:
> + rename the file

What about this?

dgw reviewed (commented):
I've had another thought since we talked about this, regarding the new module name. What if we called it instead `sopel.lifecycle`?

At the very least, that makes more sense as a title for the doc page, and it means we could have other API lifecycle stuff in there later. Fancy a decorator for alerting users to planned function signature changes, for example? (That's a quick idea, so don't judge it too hard. 😝)

If you don't like this idea, see line notes for how to improve the docs for how things are in the PR at present.

dgw commented on sopel/tools/__init__.py:
```diff
@@ -15,20 +15,15 @@
-
-from sopel import __version__
 
+from sopel.deprecated import deprecated
```
Forgot to add this before submitting the review. 😑

```suggestion
from sopel.deprecated import deprecated  # Don't delete; maintains backward compatibility with pre-8.0 API
```

dgw commented:
I'm still looking for a general reaction to calling this module `sopel.lifecycle` or something else like that.

Exirel commented:
Hm. Lifecycle could be interesting if we want to add subclass of `Warning` to be used with Python's warning system, subclasses that could be turned into exception via en env var if we wanted to (like when running test to see what must disappear).

Well, I guess it's time for me to think about that a bit more.

Exirel commented:
Rebased & squashed because conflicts, and then I moved to `sopel.lifecycle`. I don't think I can come up with a better name, and the prospect of managing version and lifecycle related things into that module is very interesting. So let's go!

dgw commented on docs/source/lifecycle.rst:
```diff
@@ -0,0 +1,31 @@
+that is deprecated: it shows when it was deprecated (Sopel 8.0), when it will
+be removed (Sopel 9.0), and how to fix that warning (use :mod:`sopel.plugin`).
+
+If this feature is primarily used by Sopel developers, Plugin authors, too, can
+take advantage of that feature with the :func:`sopel.lifecycle.deprecated`
+function documented below.
```
```suggestion
Although this feature is primarily for use by Sopel developers, plugin authors
can also take advantage of it with the :func:`sopel.lifecycle.deprecated`
function documented below.
```

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+"""Deprecation module for Sopel developers and Plugin authors.
```
I'm not a fan of German noun style, lol.

```suggestion
"""Deprecation module for Sopel developers and plugin authors.
```

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+import traceback
+from typing import Callable, Optional
+
+from pkg_resources import parse_version
```
This made me realize we don't actually have an explicit dependency on `pkg_resources` in either requirements file, and we should—_especially_ if it's used at runtime like this.

While it's _probably_ going to be available if people have `setuptools` available to install Sopel in the first place, we shouldn't implicitly rely on anything. Zen of Python, etc.

dgw replied:
Yep, to be replaced by `importlib.metadata` per my personal tickler list.

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+    """Decorator to mark deprecated functions in Sopel's API
+
+    :param reason: optional text added to the deprecation warning
+    :param version: optional version number when the decorated function
+                        is deprecated
+    :param removed_in: optional version number when the deprecated function
+                           will be removed
+    :param warning_in: optional version number when the decorated function
+                           should start emitting a warning when called
+    :param stack_frame: optional stack frame to output; defaults to
+                            ``-1``; should almost always be negative
```
```suggestion
    :param version: optional version number when the decorated function
                    is deprecated
    :param removed_in: optional version number when the deprecated function
                       will be removed
    :param warning_in: optional version number when the decorated function
                       should start emitting a warning when called
    :param stack_frame: optional stack frame to output; defaults to
                        ``-1``; should almost always be negative
```

I didn't look to see if this weird indentation was in the original `tools` definition, but… I'm gonna fix it blind.

Exirel replied:
Yeah that's weird.

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+             the decorated function, or a decorator with the appropriate
+             parameters
+
+    Any time the decorated ``func`` is called, a deprecation warning will be
+    printed to ``sys.stderr``, with the last frame of the traceback. The
```
It uses `logging` now, not `sys.stderr`.

I'm sure this is a holdover from the older implementation(s), but it's one we should fix as long as we're touching this stuff.

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+    in certain cases such as deprecating conditional behavior within an object
+    constructor, it can be useful to show a less recent stack frame instead.
+
+    .. note::
+
+        There is nothing that prevents this decorator to be used on a class's
+        method, or on any existing callable.
```
Kind of confused why this note exists.

Exirel replied:
Because when people think about decorators, they tend to think about functions only, but they can totally use that on classes and other callable objects.

dgw commented on sopel/lifecycle.py:
```diff
@@ -0,0 +1,180 @@
+    .. versionadded:: 7.1
+        The ``warning_in`` and ``stack_frame`` parameters.
+
+.. versionchanged:: 8.0
+    Moved out of :mod:`sopel.tools` to solve circular dependency issues.
```
```suggestion
    .. versionchanged:: 8.0
        Moved out of :mod:`sopel.tools` to resolve circular dependency issues.
```

dgw reviewed (changes requested):
Why does it feel like I didn't even look at the code before? Maybe you changed some of this stuff in the rename from `sopel.deprecated` to `sopel.lifecycle`, idk.

Also note, you'll want to either squash the second "renale" commit or fix that spelling error. 😸

Exirel commented:
Yup yup applied & fixed. No rebase nor squash yet, of course.

dgw reviewed (approved):
Time to squash. Really wish there was a "squash" emoji; either the food or the sport would do.
